state/state.go

```diff
@@ -230,45 +230,75 @@ func (state *HelmState) SyncReleases(helm helmexec.Interface, additionalValues [
 	return nil
 }
 
-func (state *HelmState) DiffReleases(helm helmexec.Interface, additionalValues []string) []error {
-	var wg sync.WaitGroup
+func (state *HelmState) DiffReleases(helm helmexec.Interface, additionalValues []string, workerLimit int) []error {
+	var wgRelease sync.WaitGroup
+	var wgError sync.WaitGroup
 	errs := []error{}
+	jobQueue := make(chan *ReleaseSpec, len(state.Releases))
+	errQueue := make(chan error)
 
-	for i := 0; i < len(state.Releases); i++ {
-		release := &state.Releases[i]
-		renderedName, err := renderTemplateString(release.Name)
-		if err != nil {
-			errs = append(errs, err)
-		}
-		wg.Add(1)
-		go func(wg *sync.WaitGroup, release *ReleaseSpec) {
-			// Plugin command doesn't support explicit namespace
-			release.Namespace = ""
-			flags, flagsErr := flagsForRelease(helm, state.BaseChartPath, release)
-			if flagsErr != nil {
-				errs = append(errs, flagsErr)
-			}
+	if workerLimit < 1 {
+		workerLimit = len(state.Releases)
+	}
+
+	wgRelease.Add(len(state.Releases))
 
-			for _, value := range additionalValues {
-				valfile, err := filepath.Abs(value)
+	for w := 1; w <= workerLimit; w++ {
+		go func() {
+			for release := range jobQueue {
+				errs := []error{}
+				renderedName, err := renderTemplateString(release.Name)
 				if err != nil {
 					errs = append(errs, err)
 				}
-
-				if _, err := os.Stat(valfile); os.IsNotExist(err) {
+				// Plugin command doesn't support explicit namespace
+				release.Namespace = ""
+				flags, err := flagsForRelease(helm, state.BaseChartPath, release)
+				if err != nil {
 					errs = append(errs, err)
 				}
-				flags = append(flags, "--values", valfile)
-			}
-			if len(errs) == 0 {
-				if err := helm.DiffRelease(renderedName, normalizeChart(state.BaseChartPath, release.Chart), flags...); err != nil {
-					errs = append(errs, err)
+
+				for _, value := range additionalValues {
+					valfile, err := filepath.Abs(value)
+					if err != nil {
+						errs = append(errs, err)
+					}
+
+					if _, err := os.Stat(valfile); os.IsNotExist(err) {
+						errs = append(errs, err)
+					}
+					flags = append(flags, "--values", valfile)
+				}
+
+				if len(errs) == 0 {
+					if err := helm.DiffRelease(renderedName, normalizeChart(state.BaseChartPath, release.Chart), flags...); err != nil {
+						errs = append(errs, err)
+					}
+				}
+				for _, err := range errs {
+					errQueue <- err
 				}
+				wgRelease.Done()
 			}
-			wg.Done()
-		}(&wg, release)
+		}()
 	}
-	wg.Wait()
+	wgError.Add(1)
+	go func() {
+		for err := range errQueue {
```

Comment:
Sorry if I missed something here but how this doesn't end up waiting for infinite num of errors?
I see `close(errQueue)` nowhere. When does `errQueue` gets closed so that this range loop finishes?

Reply:
Hi, not a problem. Have a look at the next few lines. The loop over the errQueue happens in a goroutine. The part beginning on line 297 first closes and waits for the release jobs and then does the same for the error jobs.

Reply:
Thx for the clarification! You are certainly correct. I somehow missed that.